portfolio/src/main/webapp/game_script.js

```diff
@@ -0,0 +1,199 @@
+      gameArea.x = e.pageX;
+      gameArea.y = e.pageY;
+    })
+  },
+  clear : function() {
+      this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
+  },
+  stop : function() {
+    clearInterval(this.interval);
+  },
+  updateInstructions : function() {
+      INSTR_UP.update();
+      INSTR_DOWN.update();
+      INSTR_MOUSE.update();
+  }
+} // gameArea
+
+/* Returns true if the current frame number
+   corresponds with the given interval */
+function everyInterval(givenInterval) {
+  if ((gameArea.frameNo / givenInterval) % 1 == 0) {return true;}
+  return false;
+}
+
+var player;
+var obstacles = [];
+var score;
+
+/* multiple intruction boxes becaise
+   canvas does not support multi-line */
+var INSTR_UP;
+var INSTR_DOWN;
+var INSTR_MOUSE;
+
+function startGame() {
+  player = new component(30, 30, "red", 10, window.innerHeight/2);
+  score = new component("30px", "Consolas", "black", window.innerWidth - 200, 30, "text");
+
+  // create a component for each instruction box
+  INSTR_UP = new component("20px", "Consolas", "black", window.innerWidth - 250, window.innerHeight - 90, "text");
+  INSTR_DOWN = new component("20px", "Consolas", "black", window.innerWidth - 250, window.innerHeight - 65, "text");
+  INSTR_MOUSE = new component("20px", "Consolas", "black", window.innerWidth - 250, window.innerHeight - 40, "text");
+
+  // unicodes for down and up arrows
+  UPArrSymbol = "\u2B06";
+  DOWNArrSymbol = "\u2B07";
+
+  INSTR_UP.text = "w/" + UPArrSymbol + " > up";
+  INSTR_DOWN.text = "s/" + DOWNArrSymbol + " > down";
+  INSTR_MOUSE.text = "m > on/off mouse control";
+
+  gameArea.start();
+}
+
+/* Construct a component with specifiec width
+   height, color and coordinates where it is
+   places in the canvas */
+function component(width, height, color, x, y, type) {
+    this.type = type;
+    this.width = width;
+    this.height = height;
+    this.speedX = 0;
+    this.speedY = 0;
+    this.x = x;
+    this.y = y;
+    this.mouseControl = false;
+    this.update = function() {
+      ctx = gameArea.context;
+      if (this.type == "text") {
+        ctx.font = this.width + " " + this.height;
+        ctx.fillStyle = color;
+        ctx.fillText(this.text, this.x, this.y);
+      } else {
+        ctx.fillStyle = color;
+        ctx.fillRect(this.x, this.y, this.width, this.height);
+      }
+    }
+    this.newPos = function() {
+        this.x += this.speedX;
+        this.y += this.speedY;
+
+        if (gameArea.x && gameArea.y && gameArea.mouseControl) {
+            /* if mouse position exists and mouse control 
+               is enabled set player y coord to mouse y coord
+               - player only moves vertically*/
+            this.y = gameArea.y;
+        }
+    }
+    this.newSpeed = function() {
+        this.speedX = 0;
+        this.speedY = 0;
+
+        keyPressed = getKey(gameArea.key);
+        switch(keyPressed) {
+            case "UP":
```

Comment:
nit: this is great for readability as you abstracted the keycodes into readable strings, and it's very easy to see what happens in each case. You could further improve this by using an enum (or enum-like construction, because JS doesn't natively support enums AFAIK), like this:

```
// somewhere outside of this function
const keyActions = {
  UP: "up"
  DOWN: "down"
}

// the switch statement
switch(keyPressed) {
  case keyActions.UP:
    // ...
```

And you can make the `getKey` function return keyActions.UP or keyActions.DOWN. 

This lets you have one place where you can define all the actions (makes the code easier to read, and you can document each value in that definition), and usually in strong-typed languages gives you extra safety. In JS you won't get this extra type safety, and if I understand correctly, you won't get any errors if you query for an action that doesn't exist (e.g. `keyActions.left`). But enums are meant for this type of data (a set of values that have a name), so it's a good idea to use them when you can.